File: 1022-Sum-of-Root-To-Leaf-Binary-Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        opened = [(root, "")]
        res = 0
        while len(opened) > 0:
            cur_node, cur_bit = opened.pop(0)
            if cur_node.left is None and cur_node.right is None:
                res += int(cur_bit + str(cur_node.val), 2)
                continue

            if cur_node.right is not None:
                opened.insert(0, (cur_node.right, cur_bit + str(cur_node.val)))
            if cur_node.left is not None:
                opened.insert(0, (cur_node.left, cur_bit + str(cur_node.val)))

        return res


# Performance Result:
#   Coding Time: 00:04:42
#   Time Complexity: O(N)
#   Space Complexity: O(N)
#   Runtime: 36 ms, faster than 73.54%
#   Memory Usage: 14.4 MB, less than 100.00%

File: 1022-Sum-of-Root-To-Leaf-Binary-Numbers.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type NextNode struct {
    Total int
    Node *TreeNode
}

func sumRootToLeaf(root *TreeNode) int {
    nextNodes := []NextNode{{Total: root.Val, Node: root}}
    res := 0
    for len(nextNodes) > 0 {
        node := nextNodes[0]
        nextNodes = nextNodes[1:]

        if node.Node.Left == nil && node.Node.Right == nil {
            res = res + node.Total
        }

        if node.Node.Left != nil {
            nextNodes = append(nextNodes,
                NextNode{Total: node.Total * 2 + node.Node.Left.Val, Node: node.Node.Left})
        }
        if node.Node.Right != nil {
            nextNodes = append(nextNodes,
                NextNode{Total: node.Total * 2 + node.Node.Right.Val, Node: node.Node.Right})
        }
    }

    return res
}


// Performance Result:
//   Coding Time: 00:29:19
//   Time Complexity: O(log n)
//   Space Complexity: O(log n)
//   Runtime: 3 ms, faster than 70.49%
//   Memory Usage: 3.50 MB, less than 6.56%
